File: app/src/main/java/ru/gidravpn/hydra/data/subscription/SingBoxConfigBuilder.kt
package ru.gidravpn.hydra.data.subscription

import ru.gidravpn.hydra.data.model.DnsEndpoint
import ru.gidravpn.hydra.data.model.GeoRoutingMode
import ru.gidravpn.hydra.data.model.HotspotSettings
import ru.gidravpn.hydra.data.model.MtuPreset
import ru.gidravpn.hydra.data.model.TlsFragmentMode
import ru.gidravpn.hydra.data.model.Protocol
import ru.gidravpn.hydra.data.model.ServerProfile
import ru.gidravpn.hydra.data.model.SplitTunnel
import ru.gidravpn.hydra.data.model.SplitTunnelMode
import org.json.JSONArray
import org.json.JSONObject

/**
 * Преобразует [ServerProfile] в полноценный конфиг sing-box (JSON), который
 * скармливается движку libbox. Собирает tun-inbound + один proxy-outbound +
 * маршрутизацию. Формат — схема sing-box 1.12 (новый формат DNS-серверов
 * с полем type; tun без auto_route — маршруты задаёт VpnService.Builder).
 */
object SingBoxConfigBuilder {

    /** Базы одной страны: IP есть всегда, домены — только у некоторых (см. GeoAssets). */
    data class GeoCountry(val code: String, val geoipPath: String, val geositePath: String? = null)

    /**
     * geoip/geosite-маршрутизация (Фаза 6c). null/OFF/пустой список — правило
     * rule_set не добавляется вовсе, поведение как раньше.
     */
    data class GeoRouting(val mode: GeoRoutingMode, val countries: List<GeoCountry>) {
        val active get() = mode != GeoRoutingMode.OFF && countries.isNotEmpty()
    }

    fun build(
        profile: ServerProfile,
        socksPort: Int = 0,
        splitTunnel: SplitTunnel = SplitTunnel(),
        dns: DnsEndpoint? = DnsEndpoint.doh("1.1.1.1"),
        geoRouting: GeoRouting? = null,
        mtu: Int = MtuPreset.AUTO.value,
        tlsFragment: TlsFragmentMode = TlsFragmentMode.OFF,
        hotspot: HotspotSettings? = null,
        ipv6: Boolean = false,
    ): JSONObject = baseConfig(outboundFor(profile, tlsFragment), splitTunnel, dns, geoRouting, mtu, hotspot, ipv6)

    /**
     * Мост Xray → tun (см. `XrayCore` в native-flavor): Xray сам tun не
     * обслуживает, поэтому он поднимается headless с локальным socks5-inbound
     * (`127.0.0.1:$socksPort`), а весь TUN/статистику/split tunneling берёт на
     * себя sing-box — единственный outbound здесь указывает на этот же порт.
     */
    fun buildXrayBridge(
        socksPort: Int,
        splitTunnel: SplitTunnel = SplitTunnel(),
        dns: DnsEndpoint? = DnsEndpoint.doh("1.1.1.1"),
        geoRouting: GeoRouting? = null,
        mtu: Int = MtuPreset.AUTO.value,
        hotspot: HotspotSettings? = null,
        ipv6: Boolean = false,
    ): JSONObject {
        val outbound = JSONObject().put("type", "socks").put("tag", "proxy")
            .put("server", "127.0.0.1").put("server_port", socksPort)
        return baseConfig(outbound, splitTunnel, dns, geoRouting, mtu, hotspot, ipv6)
    }

    private fun baseConfig(
        outbound: JSONObject,
        splitTunnel: SplitTunnel,
        dns: DnsEndpoint?,
        geoRouting: GeoRouting?,
        mtu: Int,
        hotspot: HotspotSettings?,
        ipv6: Boolean = false,
    ): JSONObject {
        val root = JSONObject()

        root.put("log", JSONObject().put("level", "info").put("timestamp", true))

        // Счётчики трафика libbox отдаёт из clash-сервера: без этого блока
        // StatusMessage.trafficAvailable = false и в UI вечные 0,0 MB.
        // external_controller не задаём — HTTP-слушатель наружу не нужен,
        // менеджер трафика работает и без него.
        root.put("experimental", JSONObject().put("clash_api", JSONObject()))

        // DNS (схема 1.12: серверы с явным type). Сюда попадают DNS-запросы
        // приложений — их перехватывает правило hijack-dns ниже. dns == null
        // (пресет "Системный резолвер") — своего сервера нет, остаётся только
        // резолвер платформы (запросы идут мимо туннеля).
        root.put("dns", JSONObject().apply {
            put("servers", JSONArray().apply {
                if (dns != null) {
                    put(JSONObject().apply {
                        put("type", dns.type); put("tag", "remote"); put("server", dns.host)
                        dns.port?.let { put("server_port", it) }
                        dns.path?.let { put("path", it) }
                        put("detour", "proxy")
                        // Хост DoH/DoT (например приватный dns.hydravpn.us) sing-box
                        // 1.12 без domain_resolver не принимает вовсе. Резолвим его
                        // через публичный DoH по IP и тоже через прокси — чтобы
                        // провайдер не видел даже имя своего DNS-сервера.
                        if (!dns.isIp) put("domain_resolver", "bootstrap")
                    })
                    if (!dns.isIp) {
                        put(JSONObject().put("type", "https").put("tag", "bootstrap")
                            .put("server", "1.1.1.1").put("detour", "proxy"))
                    }
                }
                put(JSONObject().put("type", "local").put("tag", "local"))
            })
            put("final", if (dns != null) "remote" else "local")
            put("strategy", "prefer_ipv4")
        })

        // inbound: tun (пакеты берёт наш VpnService через файловый дескриптор;
        // auto_route/strict_route выключены — маршрутизацией владеет VpnService.Builder)
        root.put("inbounds", JSONArray().apply {
            put(JSONObject().apply {
                put("type", "tun")
                put("tag", "tun-in")
                put("interface_name", "hydra-tun")
                put("mtu", mtu)
                // IPv6-адрес — только в режиме Ipv6Mode.ENABLE; тот же, что у VpnService.Builder.
                put("address", JSONArray().put("172.19.0.1/28").apply {
                    if (ipv6) put("${ru.gidravpn.hydra.data.model.Ipv6Mode.TUN_ADDRESS}/${ru.gidravpn.hydra.data.model.Ipv6Mode.TUN_PREFIX}")
                })
                put("auto_route", false)
                put("strict_route", false)
                put("stack", "gvisor")
            })
            // Хотспот-прокси (6f): SOCKS+HTTP на всех интерфейсах, только с логином/
            // паролем. Идёт через те же route.rules, что и tun: sniff, DNS, split, geo.
            if (hotspot != null && hotspot.isUsable) {
                put(JSONObject().apply {
                    put("type", "mixed")
                    put("tag", "hotspot-in")
                    put("listen", "0.0.0.0")
                    put("listen_port", hotspot.port)
                    put("users", JSONArray().put(
                        JSONObject().put("username", hotspot.username).put("password", hotspot.password)
                    ))
                })
            }
        })

        // outbounds: proxy + direct. Спецаутбаунды dns/block (устарели в 1.11,
        // удаляются в 1.13) заменены rule actions — см. sniff/hijack-dns ниже.
        root.put("outbounds", JSONArray().apply {
            put(outbound)
            put(JSONObject().put("type", "direct").put("tag", "direct"))
        })

        // маршрутизация (+ пользовательские правила split tunneling по IP/доменам — Фаза 2)
        val netRules = splitTunnel.netRules.takeIf { splitTunnel.netActive }.orEmpty()
        val geo = geoRouting?.takeIf { it.active }
        // Теги в том же порядке, что и rule_set: geoip-<cc>, затем geosite-<cc>.
        val geoTags = geo?.countries.orEmpty().flatMap { c ->
            listOfNotNull("geoip-${c.code}" to c.geoipPath, c.geositePath?.let { "geosite-${c.code}" to it })
        }
        root.put("route", JSONObject().apply {
            if (geo != null) {
                put("rule_set", JSONArray().apply {
                    geoTags.forEach { (tag, path) ->
                        put(JSONObject().put("type", "local").put("tag", tag)
                            .put("format", "binary").put("path", path))
                    }
                })
            }
            put("rules", JSONArray().apply {
                // sniff обязан идти первым: без него у соединения нет ни протокола,
                // ни домена. Раньше его не было вовсе — `protocol: dns` не срабатывал
                // (DNS уходил в прокси сырым UDP мимо dns-серверов выше), а доменные
                // правила (netRules DOMAIN*, geosite-ru) не матчились никогда.
                put(JSONObject().put("action", "sniff"))
                put(JSONObject().put("protocol", "dns").put("action", "hijack-dns"))
                if (netRules.isNotEmpty()) {
                    val ruleOutbound = if (splitTunnel.netMode == SplitTunnelMode.EXCLUDE) "direct" else "proxy"
                    netRules.groupBy({ it.type }, { it.value }).forEach { (type, values) ->
                        put(JSONObject().put(type.singBoxKey, JSONArray(values)).put("outbound", ruleOutbound))
                    }
                }
                put(JSONObject().put("ip_is_private", true).put("outbound", "direct"))
                // Любой из rule_set выбранных стран (OR внутри одного правила —
                // стандартная семантика sing-box для массива в одном поле): DIRECT
                // ведёт их мимо VPN, VIA_PROXY — наоборот, единственное, что идёт в прокси.
                if (geo != null) {
                    val geoOutbound = if (geo.mode == GeoRoutingMode.DIRECT) "direct" else "proxy"
                    put(JSONObject()
                        .put("rule_set", JSONArray(geoTags.map { it.first }))
                        .put("outbound", geoOutbound))
                }
            })
            // Явный whitelist по IP/доменам (INCLUDE) — сильнее geo-режима: то, что
            // не попало под явные правила, идёт мимо VPN независимо от geoRouting.
            val netIncludeActive = netRules.isNotEmpty() && splitTunnel.netMode == SplitTunnelMode.INCLUDE
            put("final", when {
                netIncludeActive -> "direct"
                geo?.mode == GeoRoutingMode.VIA_PROXY -> "direct"
                else -> "proxy"
            })
            put("auto_detect_interface", true)
            // Домен самого прокси-сервера резолвим платформенным резолвером, мимо
            // туннеля и мимо dns.final ("remote" идёт через этот же прокси —
            // резолвить его адрес через него самого = замкнутый круг). Без этого
            // поля sing-box 1.12 предупреждает, а в 1.14 откажется стартовать.
            put("default_domain_resolver", "local")
        })

        return root
    }

    private fun outboundFor(p: ServerProfile, tlsFragment: TlsFragmentMode = TlsFragmentMode.OFF): JSONObject {
        val o = JSONObject().put("tag", "proxy").put("server", p.address).put("server_port", p.port)
        val extra = runCatching { JSONObject(p.extra) }.getOrDefault(JSONObject())

        when (p.protocol) {
            Protocol.VLESS -> {
                o.put("type", "vless").put("uuid", p.uuidOrPassword)
                if (p.flow.isNotEmpty()) o.put("flow", p.flow)
                o.put("tls", tlsBlock(p, extra, tlsFragment))
                transportBlock(p)?.let { o.put("transport", it) }
            }
            Protocol.VMESS -> {
                o.put("type", "vmess").put("uuid", p.uuidOrPassword)
                    .put("security", "auto").put("alter_id", extra.optInt("aid", 0))
                if (p.security == "tls") o.put("tls", tlsBlock(p, extra, tlsFragment))
                transportBlock(p)?.let { o.put("transport", it) }
            }
            Protocol.TROJAN -> {
                o.put("type", "trojan").put("password", p.uuidOrPassword)
                o.put("tls", tlsBlock(p, extra, tlsFragment))
                transportBlock(p)?.let { o.put("transport", it) }
            }
            Protocol.SHADOWSOCKS -> {
                o.put("type", "shadowsocks")
                    .put("method", extra.optString("method", "aes-256-gcm"))
                    .put("password", p.uuidOrPassword)
            }
            Protocol.HYSTERIA2 -> {
                o.put("type", "hysteria2").put("password", p.uuidOrPassword)
                o.put("tls", tlsBlock(p, extra))
                if (extra.has("obfs")) o.put("obfs", JSONObject()
                    .put("type", "salamander").put("password", extra.optString("obfs_password")))
            }
            Protocol.WIREGUARD -> {
                // Обычный WireGuard через sing-box (для AmneziaWG — отдельный движок amneziawg-go)
                o.put("type", "wireguard")
                    .put("private_key", p.uuidOrPassword)
                    .put("peer_public_key", extra.optString("public_key"))
                    .put("local_address", JSONArray(
                        extra.optString("local_address").ifBlank { "172.19.0.2/32" }
                            .split(",").map { it.trim() }
                    ))
                if (extra.has("preshared_key")) o.put("pre_shared_key", extra.optString("preshared_key"))
                if (extra.has("mtu")) o.put("mtu", extra.optInt("mtu", 1408))
            }
            Protocol.TUIC -> {
                o.put("type", "tuic").put("uuid", p.uuidOrPassword)
                    .put("password", extra.optString("password"))
                    .put("congestion_control", extra.optString("congestion_control", "bbr"))
                o.put("tls", tlsBlock(p, extra))
            }
            else -> o.put("type", "direct")  // SSTP/L2TP/PPTP/AWG обрабатываются отдельными движками, не sing-box
        }
        return o
    }

    private fun tlsBlock(
        p: ServerProfile,
        extra: JSONObject,
        fragment: TlsFragmentMode = TlsFragmentMode.OFF,
    ): JSONObject {
        val tls = JSONObject().put("enabled", true)
            .put("server_name", p.sni.ifBlank { p.address })
        if (p.alpn.isNotBlank()) tls.put("alpn", JSONArray(p.alpn.split(",").map { it.trim() }))
        // uTLS отпечаток
        tls.put("utls", JSONObject().put("enabled", true).put("fingerprint", p.fingerprint))
        // REALITY
        if (p.security == "reality" && extra.has("reality_pbk")) {
            tls.put("reality", JSONObject()
                .put("enabled", true)
                .put("public_key", extra.optString("reality_pbk"))
                .put("short_id", extra.optString("reality_sid")))
        }
        when (fragment) {
            TlsFragmentMode.OFF -> Unit
            TlsFragmentMode.RECORD -> tls.put("record_fragment", true)
            TlsFragmentMode.TCP -> tls.put("fragment", true)
        }
        return tls
    }

    private fun transportBlock(p: ServerProfile): JSONObject? = when (p.transport) {
        "ws" -> JSONObject().put("type", "ws")
            .put("path", p.transportPath.ifBlank { "/" })
        "grpc" -> JSONObject().put("type", "grpc")
            .put("service_name", p.transportPath)
        "http" -> JSONObject().put("type", "http").put("path", p.transportPath.ifBlank { "/" })
        else -> null   // tcp — транспорт не указывается
    }
}

File: ios/Packages/HydraKit/Sources/HydraKit/SingBoxConfigBuilder.swift
import Foundation

/// Конфиг sing-box 1.12 для iOS — порт `SingBoxConfigBuilder` с Android. Правила маршрутизации,
/// DNS, outbound-ы, фрагментация TLS — один в один. Отличие только в tun: на Android маршрутами
/// владеет VpnService.Builder (`auto_route: false`), а на iOS sing-box сам сообщает их системе через
/// `openTun` → NEPacketTunnelNetworkSettings, поэтому `auto_route: true`.
public enum SingBoxConfigBuilder {

    public struct GeoCountry: Sendable {
        public var code: String
        public var geoipPath: String
        public var geositePath: String?
        public init(code: String, geoipPath: String, geositePath: String? = nil) {
            self.code = code; self.geoipPath = geoipPath; self.geositePath = geositePath
        }
    }

    public struct Options: Sendable {
        public var routing = RoutingSettings()
        public var geo: [GeoCountry] = []
        public var hotspot: (port: Int, user: String, password: String)?
        public init() {}
    }

    public static func build(_ profile: ServerProfile, _ o: Options) throws -> [String: Any] {
        var root: [String: Any] = [:]
        root["log"] = ["level": "info", "timestamp": true]
        root["experimental"] = ["clash_api": [String: Any]()]

        let dns = o.routing.resolvedDns()
        var servers: [[String: Any]] = []
        if let dns {
            var s: [String: Any] = ["type": dns.type, "tag": "remote", "server": dns.host, "detour": "proxy"]
            if let p = dns.port { s["server_port"] = p }
            if let p = dns.path { s["path"] = p }
            if !dns.isIp { s["domain_resolver"] = "bootstrap" }
            servers.append(s)
            if !dns.isIp {
                servers.append(["type": "https", "tag": "bootstrap", "server": "1.1.1.1", "detour": "proxy"])
            }
        }
        servers.append(["type": "local", "tag": "local"])
        // IPv6 «Блокировать»: DNS отдаёт только A-записи — приложения сразу идут по IPv4.
        let v4only = o.routing.ipv6 == .block
        root["dns"] = ["servers": servers, "final": dns != nil ? "remote" : "local",
                       "strategy": v4only ? "ipv4_only" : "prefer_ipv4"]

        // IPv6-адрес у tun есть всегда: без него система не отдаёт туннелю IPv6-маршруты, и IPv6
        // шёл бы мимо VPN (а NEIPv6Settings без адресов iOS не принимает — трюк Android с «::/0 без
        // адреса» здесь не работает). В режиме «Блокировать» IPv6 отклоняется правилом ниже.
        let tun: [String: Any] = [
            "type": "tun", "tag": "tun-in", "mtu": o.routing.mtu.value,
            "address": ["172.19.0.1/28", "fdfe:dcba:9876::1/126"],
            "auto_route": true, "strict_route": false, "stack": "gvisor",
        ]
        var inbounds: [[String: Any]] = [tun]
        if let h = o.hotspot, !h.password.isEmpty {
            inbounds.append([
                "type": "mixed", "tag": "hotspot-in", "listen": "0.0.0.0", "listen_port": h.port,
                "users": [["username": h.user, "password": h.password]],
            ])
        }
        root["inbounds"] = inbounds
        root["outbounds"] = [try outbound(profile, o.routing.tlsFragment), ["type": "direct", "tag": "direct"]]

        let netRules = o.routing.netActive ? o.routing.netRules : []
        let geoActive = o.routing.geoMode != .off && !o.geo.isEmpty
        var geoTags: [(String, String)] = []
        if geoActive {
            for c in o.geo {
                geoTags.append(("geoip-\(c.code)", c.geoipPath))
                if let s = c.geositePath { geoTags.append(("geosite-\(c.code)", s)) }
            }
        }
        var route: [String: Any] = [:]
        if geoActive {
            route["rule_set"] = geoTags.map { ["type": "local", "tag": $0.0, "format": "binary", "path": $0.1] }
        }
        var rules: [[String: Any]] = [["action": "sniff"], ["protocol": "dns", "action": "hijack-dns"]]
        if v4only { rules.append(["ip_version": 6, "action": "reject"]) }
        if !netRules.isEmpty {
            let target = o.routing.netMode == .exclude ? "direct" : "proxy"
            var grouped: [(NetRuleType, [String])] = []
            for r in netRules {
                if let i = grouped.firstIndex(where: { $0.0 == r.type }) { grouped[i].1.append(r.value) }
                else { grouped.append((r.type, [r.value])) }
            }
            for (type, values) in grouped { rules.append([type.singBoxKey: values, "outbound": target]) }
        }
        rules.append(["ip_is_private": true, "outbound": "direct"])
        if geoActive {
            rules.append(["rule_set": geoTags.map(\.0), "outbound": o.routing.geoMode == .direct ? "direct" : "proxy"])
        }
        route["rules"] = rules
        let includeOnly = !netRules.isEmpty && o.routing.netMode == .include
        route["final"] = (includeOnly || (geoActive && o.routing.geoMode == .viaProxy)) ? "direct" : "proxy"
        route["auto_detect_interface"] = true
        route["default_domain_resolver"] = "local"
        root["route"] = route
        return root
    }

    public static func buildJSON(_ profile: ServerProfile, _ o: Options) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: try build(profile, o), options: [.prettyPrinted, .sortedKeys])
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    public enum BuildError: Error, LocalizedError {
        case unsupported(String)
        public var errorDescription: String? {
            switch self { case .unsupported(let p): "\(p): протокол пока не поддерживается на iOS" }
        }
    }

    static func outbound(_ p: ServerProfile, _ fragment: TlsFragmentMode) throws -> [String: Any] {
        var o: [String: Any] = ["tag": "proxy", "server": p.address, "server_port": p.port]
        let extra = p.extraObject
        func str(_ k: String) -> String? {
            if let s = extra[k] as? String { return s }
            if let n = extra[k] as? NSNumber { return n.stringValue }
            return nil
        }
        switch p.serverProtocol {
        case .vless:
            o["type"] = "vless"; o["uuid"] = p.uuidOrPassword
            if !p.flow.isEmpty { o["flow"] = p.flow }
            o["tls"] = tls(p, extra, fragment)
            if let t = transport(p) { o["transport"] = t }
        case .vmess:
            o["type"] = "vmess"; o["uuid"] = p.uuidOrPassword; o["security"] = "auto"
            o["alter_id"] = Int(str("aid") ?? "0") ?? 0
            if p.security == "tls" { o["tls"] = tls(p, extra, fragment) }
            if let t = transport(p) { o["transport"] = t }
        case .trojan:
            o["type"] = "trojan"; o["password"] = p.uuidOrPassword
            o["tls"] = tls(p, extra, fragment)
            if let t = transport(p) { o["transport"] = t }
        case .shadowsocks:
            o["type"] = "shadowsocks"; o["method"] = str("method") ?? "aes-256-gcm"; o["password"] = p.uuidOrPassword
        case .hysteria2:
            o["type"] = "hysteria2"; o["password"] = p.uuidOrPassword
            o["tls"] = tls(p, extra, .off)
            if extra["obfs"] != nil { o["obfs"] = ["type": "salamander", "password": str("obfs_password") ?? ""] }
        case .wireguard:
            o["type"] = "wireguard"; o["private_key"] = p.uuidOrPassword
            o["peer_public_key"] = str("public_key") ?? ""
            let local = (str("local_address") ?? "").isEmpty ? "172.19.0.2/32" : str("local_address")!
            o["local_address"] = local.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
            if let psk = str("preshared_key") { o["pre_shared_key"] = psk }
            if let m = str("mtu"), let mtu = Int(m) { o["mtu"] = mtu }
        case .tuic:
            o["type"] = "tuic"; o["uuid"] = p.uuidOrPassword; o["password"] = str("password") ?? ""
            o["congestion_control"] = str("congestion_control") ?? "bbr"
            o["tls"] = tls(p, extra, .off)
        default:
            throw BuildError.unsupported(p.serverProtocol?.displayName ?? p.protocolId)
        }
        return o
    }

    static func tls(_ p: ServerProfile, _ extra: [String: Any], _ fragment: TlsFragmentMode) -> [String: Any] {
        var t: [String: Any] = ["enabled": true, "server_name": p.sni.isEmpty ? p.address : p.sni]
        if !p.alpn.isEmpty { t["alpn"] = p.alpn.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) } }
        t["utls"] = ["enabled": true, "fingerprint": p.fingerprint]
        if p.security == "reality", let pbk = extra["reality_pbk"] as? String {
            t["reality"] = ["enabled": true, "public_key": pbk, "short_id": (extra["reality_sid"] as? String) ?? ""]
        }
        switch fragment {
        case .off: break
        case .record: t["record_fragment"] = true
        case .tcp: t["fragment"] = true
        }
        return t
    }

    static func transport(_ p: ServerProfile) -> [String: Any]? {
        switch p.transport {
        case "ws": return ["type": "ws", "path": p.transportPath.isEmpty ? "/" : p.transportPath]
        case "grpc": return ["type": "grpc", "service_name": p.transportPath]
        case "http": return ["type": "http", "path": p.transportPath.isEmpty ? "/" : p.transportPath]
        default: return nil
        }
    }
}
